File: app/src/main/java/io/treehouses/remote/bases/BaseServicesFragment.java
package io.treehouses.remote.bases;

import android.app.AlertDialog;
import android.content.Intent;
import android.net.Uri;
import android.text.SpannableString;
import android.text.method.LinkMovementMethod;
import android.text.util.Linkify;
import android.util.Log;
import android.widget.TextView;
import android.widget.Toast;

import com.google.gson.Gson;

import org.json.JSONException;
import org.json.JSONObject;

import java.util.ArrayList;
import java.util.Collections;

import io.treehouses.remote.pojo.ServiceInfo;
import io.treehouses.remote.pojo.ServicesData;

public class BaseServicesFragment extends BaseFragment {
    private static final int[] MINIMUM_VERSION = {1, 14, 1};
    private String startJson = "";
    private boolean gettingJSON = false;
    protected ServicesData servicesData;

    protected void openLocalURL(String url) {
        Intent intent = new Intent(Intent.ACTION_VIEW, Uri.parse("http://" + url));

        Log.d("OPENING: ", "http://" + url + "||");
        String title = "Select a browser";
        Intent chooser = Intent.createChooser(intent, title);

        if (intent.resolveActivity(getContext().getPackageManager()) != null) startActivity(chooser);
    }

    protected void openTorURL (String url) {
        Intent intent = getContext().getPackageManager().getLaunchIntentForPackage("org.torproject.torbrowser");
        if (intent != null) {
            intent.setData(Uri.parse("http://" + url));
            intent.setAction(Intent.ACTION_VIEW);
            startActivity(intent);
        }
        else {
            final String s = "Please install Tor Browser from: \n\n https://play.google.com/store/apps/details?id=org.torproject.torbrowser";
            final SpannableString spannableString = new SpannableString(s);
            Linkify.addLinks(spannableString, Linkify.ALL);
            AlertDialog alertDialog = new AlertDialog.Builder(getContext()).setTitle("Tor Browser Not Found").setMessage(spannableString).create();
            alertDialog.show();
            TextView alertTextView = (TextView) alertDialog.findViewById(android.R.id.message);
            alertTextView.setMovementMethod(LinkMovementMethod.getInstance());
        }
    }

    protected boolean checkVersion(int[] versionIntNumber) {
        if (versionIntNumber[0] > MINIMUM_VERSION[0]) return true;
        if (versionIntNumber[0] == MINIMUM_VERSION[0] && versionIntNumber[1] > MINIMUM_VERSION[1]) return true;
        return (versionIntNumber[0] == MINIMUM_VERSION[0]) && (versionIntNumber[1] == MINIMUM_VERSION[1]) && (versionIntNumber[2] >= MINIMUM_VERSION[2]);
    }
    protected void writeToRPI(String ping) {
        mChatService.write(ping.getBytes());
    }

    protected void performService(String action, String command, String name) {
        Log.d("SERVICES", action + " " + name);
        Toast.makeText(getContext(), name + " " + action, Toast.LENGTH_LONG).show();
        writeToRPI(command);
    }

    protected boolean installedOrRunning(ServiceInfo selected) {
        return selected.serviceStatus == ServiceInfo.SERVICE_INSTALLED || selected.serviceStatus == ServiceInfo.SERVICE_RUNNING;
    }

    private void constructServiceList(ServicesData servicesData, ArrayList<ServiceInfo> services) {
        if (servicesData == null || servicesData.getAvailable() == null) {
            Toast.makeText(getContext(), "Error Occurred. Please Refresh", Toast.LENGTH_SHORT).show();
            return;
        }
        services.clear();

        addServicesToList(services);
        getServices(services);

        formatList(services);
    }

    private void addServicesToList( ArrayList<ServiceInfo> services){
        for (String service : servicesData.getAvailable()) {
            if (inServiceList(service, services) == -1) {
                services.add(new ServiceInfo(service, ServiceInfo.SERVICE_AVAILABLE, servicesData.getIcon().get(service),
                        servicesData.getInfo().get(service), servicesData.getAutorun().get(service)));
            }
        }
    }

    private void getServices(ArrayList<ServiceInfo> services){
        for (String service : servicesData.getRunning()) {
            if (inServiceList(service, services) == -1) continue;
                runningOrInstalled(service, services, true);
        }
        for (String service : servicesData.getRunning()) {
            if (inServiceList(service, services) == -1) continue;
                runningOrInstalled(service, services, false);
        }
    }

    private void runningOrInstalled(String service, ArrayList<ServiceInfo> services, boolean installedOrRunning){
        if(installedOrRunning)
            services.get(inServiceList(service, services)).serviceStatus = ServiceInfo.SERVICE_INSTALLED;
        else
            services.get(inServiceList(service, services)).serviceStatus = ServiceInfo.SERVICE_RUNNING;
    }

    private void formatList(ArrayList<ServiceInfo> services) {
        if (inServiceList("Installed", services) == -1) services.add(0, new ServiceInfo("Installed", ServiceInfo.SERVICE_HEADER_INSTALLED));
        if (inServiceList("Available", services) == -1) services.add(0, new ServiceInfo("Available",ServiceInfo.SERVICE_HEADER_AVAILABLE));
        Collections.sort(services);
    }

    protected int inServiceList(String name, ArrayList<ServiceInfo> services) {
        for (int i = 0; i < services.size(); i++) {
            if (services.get(i).name.equals(name)) return i;
        }
        return -1;
    }

    protected boolean isVersionNumber(String s, int[] versionNumber) {
        if (!s.contains(".")) return false;
        String[] parts = s.split("[.]");
        int[] intParts = new int[3];
        if (parts.length != 3) return false;
        for (int i = 0; i < parts.length; i++) {
            try {
                intParts[i] = Integer.parseInt(parts[i].trim());
            } catch (NumberFormatException e) {
                return false;
            }
        }
        System.arraycopy(intParts,0, versionNumber, 0, 3);
        return true;
    }

    private int isError(String output) {
        if(output.toLowerCase().startsWith("usage:") ||
                output.toLowerCase().contains("error") ||
                output.toLowerCase().contains("unknown"))
            return 0;
        else
            return -1;
    }

    protected int performAction(String output, ArrayList<ServiceInfo> services) {
        int i = isError(output);
        startJson += output.trim();

        if (gettingJSON && startJson.endsWith("}}") ) {
            startJson += output.trim();
            try {
                JSONObject jsonObject = new JSONObject(startJson);
                servicesData = new Gson().fromJson(jsonObject.toString(), ServicesData.class);
                constructServiceList(servicesData, services);
            } catch (JSONException e) {
                e.printStackTrace();
            }
            gettingJSON = false;
            i = 1;
        }
        else if (output.trim().startsWith("{")) {
            Log.d("STARTED", "performAction: ");
            startJson = output.trim();
            gettingJSON = true;
        }
        return i;
    }

    protected boolean isTorURL(String output, boolean received) {
        return output.contains(".onion") && !received;
    }

    protected boolean isLocalUrl(String output, boolean received) {
        return output.contains(".") && output.contains(":") && output.length() < 25 && !received;
    }
}

File: app/src/main/java/io/treehouses/remote/bases/BaseServicesFragment.kt
package io.treehouses.remote.bases

import android.app.AlertDialog
import android.content.Intent
import android.net.Uri
import android.text.SpannableString
import android.text.method.LinkMovementMethod
import android.text.util.Linkify
import android.util.Log
import android.view.ContextThemeWrapper
import android.view.View
import android.widget.TextView
import android.widget.Toast
import com.google.gson.Gson
import io.treehouses.remote.R
import io.treehouses.remote.pojo.ServiceInfo
import io.treehouses.remote.pojo.ServicesData
import io.treehouses.remote.utils.RESULTS
import io.treehouses.remote.utils.match
import org.json.JSONException
import org.json.JSONObject
import java.util.*

open class BaseServicesFragment : BaseFragment() {
    private var startJson = ""
    private var gettingJSON = false
    protected var servicesData: ServicesData? = null
    protected fun openLocalURL(url: String) {
        val intent = Intent(Intent.ACTION_VIEW, Uri.parse("http://$url"))
        Log.d("OPENING: ", "http://$url||")
        val title = "Select a browser"
        val chooser = Intent.createChooser(intent, title)
        if (intent.resolveActivity(requireContext().packageManager) != null) startActivity(chooser)
    }

    protected fun openTorURL(url: String) {
        val intent = requireContext().packageManager.getLaunchIntentForPackage("org.torproject.torbrowser")
        if (intent != null) {
            intent.data = Uri.parse("http://$url")
            intent.action = Intent.ACTION_VIEW
            startActivity(intent)
        } else {
            val s = "Please install Tor Browser from: \n\n https://play.google.com/store/apps/details?id=org.torproject.torbrowser"
            val spannableString = SpannableString(s)
            Linkify.addLinks(spannableString, Linkify.ALL)
            val alertDialog = AlertDialog.Builder(ContextThemeWrapper(context, R.style.CustomAlertDialogStyle)).setTitle("Tor Browser Not Found").setMessage(spannableString).create()
            alertDialog.show()
            val alertTextView = alertDialog.findViewById<View>(android.R.id.message) as TextView
            alertTextView.movementMethod = LinkMovementMethod.getInstance()
        }
    }

    protected fun checkVersion(versionIntNumber: IntArray): Boolean {
        if (versionIntNumber[0] > MINIMUM_VERSION[0]) return true
        return if (versionIntNumber[0] == MINIMUM_VERSION[0] && versionIntNumber[1] > MINIMUM_VERSION[1]) true else versionIntNumber[0] == MINIMUM_VERSION[0] && versionIntNumber[1] == MINIMUM_VERSION[1] && versionIntNumber[2] >= MINIMUM_VERSION[2]
    }

    protected fun writeToRPI(ping: String) {
        mChatService.write(ping.toByteArray())
    }

    protected fun performService(action: String, command: String, name: String) {
        Log.d("SERVICES", "$action $name")
        Toast.makeText(context, "$name $action", Toast.LENGTH_LONG).show()
        writeToRPI(command)
    }

    protected fun installedOrRunning(selected: ServiceInfo): Boolean {
        return selected.serviceStatus == ServiceInfo.SERVICE_INSTALLED || selected.serviceStatus == ServiceInfo.SERVICE_RUNNING
    }

    private fun constructServiceList(servicesData: ServicesData?, services: ArrayList<ServiceInfo>) {
        if (servicesData == null || servicesData.available == null) {
            Toast.makeText(context, "Error Occurred. Please Refresh", Toast.LENGTH_SHORT).show()
            return
        }
        services.clear()
        addServicesToList(services)
        getServices(services)
        formatList(services)
    }

    private fun addServicesToList(services: ArrayList<ServiceInfo>) {
        for (service in servicesData!!.available) {
            if (inServiceList(service, services) == -1) {
                services.add(ServiceInfo(service, ServiceInfo.SERVICE_AVAILABLE, servicesData!!.icon[service],
                        servicesData!!.info[service], servicesData!!.autorun[service]))
            }
        }
    }

    private fun getServices(services: ArrayList<ServiceInfo>) {
        for (service in servicesData!!.installed) {
            checkInServicesList(service, services, true)
        }
        for (service in servicesData!!.running) {
            checkInServicesList(service, services, false)
        }
    }

    private fun checkInServicesList(service: String, services: ArrayList<ServiceInfo>, installedOrRunning: Boolean) {
        if (inServiceList(service, services) != -1) installedOrRunning(service, services, installedOrRunning)
    }

    private fun installedOrRunning(service: String, services: ArrayList<ServiceInfo>, installedOrRunning: Boolean) {
        if (installedOrRunning) services[inServiceList(service, services)].serviceStatus = ServiceInfo.SERVICE_INSTALLED else services[inServiceList(service, services)].serviceStatus = ServiceInfo.SERVICE_RUNNING
    }

    private fun formatList(services: ArrayList<ServiceInfo>) {
        if (inServiceList("Installed", services) == -1) services.add(0, ServiceInfo("Installed", ServiceInfo.SERVICE_HEADER_INSTALLED))
        if (inServiceList("Available", services) == -1) services.add(0, ServiceInfo("Available", ServiceInfo.SERVICE_HEADER_AVAILABLE))
        Collections.sort(services)
    }

    protected fun inServiceList(name: String, services: ArrayList<ServiceInfo>): Int {
        for (i in services.indices) {
            if (services[i].name == name) return i
        }
        return -1
    }

    protected fun isVersionNumber(s: String, versionNumber: IntArray?): Boolean {
        if (!s.contains(".")) return false
        val parts = s.split("[.]".toRegex()).toTypedArray()
        val intParts = IntArray(3)
        if (parts.size != 3) return false
        for (i in parts.indices) {
            try {
                intParts[i] = parts[i].trim { it <= ' ' }.toInt()
            } catch (e: NumberFormatException) {
                return false
            }
        }
        System.arraycopy(intParts, 0, versionNumber, 0, 3)
        return true
    }

    private fun isError(output: String): Int {
        return if (match(output) === RESULTS.ERROR) 0 else -1
    }

    protected fun performAction(output: String, services: ArrayList<ServiceInfo>): Int {
        var i = isError(output)
        startJson += output.trim { it <= ' ' }
        if (gettingJSON && startJson.endsWith("}}")) {
            startJson += output.trim { it <= ' ' }
            try {
                val jsonObject = JSONObject(startJson)
                servicesData = Gson().fromJson(jsonObject.toString(), ServicesData::class.java)
                constructServiceList(servicesData, services)
            } catch (e: JSONException) {
                e.printStackTrace()
            }
            gettingJSON = false
            i = 1
        } else if (output.trim { it <= ' ' }.startsWith("{")) {
            Log.d("STARTED", "performAction: ")
            startJson = output.trim { it <= ' ' }
            gettingJSON = true
        }
        return i
    }

    protected fun isTorURL(output: String, received: Boolean): Boolean {
        return output.contains(".onion") && !received
    }

    protected fun isLocalUrl(output: String, received: Boolean): Boolean {
        return output.contains(".") && output.contains(":") && output.length < 25 && !received
    }

    companion object {
        private val MINIMUM_VERSION = intArrayOf(1, 14, 1)
    }
}